### ISSUE TITLE
Feature request: restart on condition

### ISSUE THREAD

**Issue by zimbatm** [mattgreen/watchexec#83]
## Problem statement

One a build & run scenario:

`./dev-start` script
```sh
#!/bin/sh -e
watchexec --restart --signal SIGKILL ./build-and-run
```

`./build-and-run` script:
```sh
#!/bin/sh -e
make
exec ./server
```

Some times `make` doesn't produce any new output but the server is still restarted. It would be nice if there was a way to avoid that.

---

### Idea 1

Extend watchexec to run the build before the main script. This would essentially separate the build and supervisor bits.

```sh
watchexec --restart-on ./build --signal SIGKILL ./server
```

Where `./build` would have different exit statuses:
* `0` success and restart
* `123` success, don't restart
* `any other`: failure, don't restart

### Current workaround

Remove the process-manager bits from watchman. In that scenario, watchman invokes a script that communicated with the process manager and implements it's own restart logic.

Example:
`./dev-start` script
```sh
#!/bin/sh -e
supervisord -c supervisor.conf
watchexec ./build-and-run
```

`./build-and-run` script
```sh
#!/bin/sh -e
server_mtime() {
  stat -c "%Y" ./server || echo 0
}

old_mtime=$(server_mtime)
make

if [ "$(server_mtime)" -gt "$old_mtime" ]; then
  supervisorctl -c supervisor.conf restart server
fi
```

Thanks for listening!

**Comment by passcod** [comment]
Going to attempt something approaching this in 1.12.0 related to #101 and #132

**Comment by calebschoepp** [comment]
Hey @passcod thanks for your work on this project.

I have a similar issue where I'm trying to build a single command [`spin watch`](https://github.com/fermyon/spin/pull/1164) that both rebuilds a project and restarts the server. I need the rebuild to occur when source files are modified but I only want the server to restart when the artifact produced by the build changes.

It sounds to me like the solution listed above of using two instances would work for me. I have couple questions about it though:
1. Could you point me to where that is happening in cargo-watch so I could read some source code.
2. If I'm running to `Watchexec` instances in a single process will there be any conflicts or race conditions I would need to worry about?
3. How much overhead do you think that is creating? I suppose we'd be doubly listening to the filesystem but I don't have a sense of how expensive that would be.
4. What will happen as both the build and server processes are competing to log to the screen? Is there any solution to make them cooperate in the logging?

**Comment by passcod** [comment]
1. go from watchexec instead. sounds like you've already found the docs for that
2. yes, signal handling. you'll want to handle signals on one instance only
3. if you're going custom you don't really need to watch the filesystem twice
4. they'll interleave

i'd go a bit deeper than making two instances and make your own from the components instead. open up the source for the default one and replicate in your source, then trim to just what you need. that way you can get two Supervisor instances instead of just one to hold onto the two processes and hit up one or the other at your leisure.

**Comment by calebschoepp** [comment]
1,2,3. Got it.
4. In the cargo-watch use case did you do anything to prevent log interleaving?

That makes a lot of sense to me. The alternative poor-man's approach I thought of was running two instances of `Watchexec` each in their own process so they don't conflict. It really would be ideal to do it in one process though so I'll give your suggestion a go.

**Comment by calebschoepp** [comment]
Somewhat unrelated to this issue: Does watchexec support restarting long running processes? I'm using the following outcome to restart my server:

```rust
action.outcome(Outcome::if_running(
    Outcome::both(Outcome::Stop, Outcome::Start),
    Outcome::Start,
));
```

I'm getting the error below. It is shutting down the server properly (not holding on to the port) but is erroring for a reason I can't understand.

```
2023-02-28T00:49:03.888231Z ERROR watchexec::command::supervisor: while waiting on process err=io(waiting on process group): blocking waitpid returned pid=0
2023-02-28T00:49:03.888385Z ERROR watchexec::watchexec: runtime error err=io(waiting on process group): blocking waitpid returned pid=0
ErrorHook { error: IoError { about: "waiting on process group", err: Custom { kind: Other, error: "blocking waitpid returned pid=0" } }, critical: OnceCell(Uninit) }
```

Note that I'm using grouped commands because my server spins up children processes.

**Comment by passcod** [comment]
Are you using the latest watchexec patch? There's a regression introduced by recent Rust that manifests like this, and it was fixed in latest.

**Comment by calebschoepp** [comment]
> Are you using the latest watchexec patch? There's a regression introduced by recent Rust that manifests like this, and it was fixed in latest.

No, I'm not. I was using a commit off of someones fork that fixed the gitx issue because I can't get watchexec to compile.

```
watchexec = { git = "https://github.com/david-a-perez/watchexec.git", rev = "10d7c7fb05495791cc14a6f0ba5164d449e7a66c"}
```

Is there a fix for #506 with the library (not the CLI)? Which commit should I be pointing to?
